### ISSUE TITLE
type / submit in Cypress not triggering HTML validation on elements

### ISSUE THREAD

**Issue by brettz9** [cypress-io/cypress#6678]
### Current behavior:

Not finding invalid elements, whether by:

1. `cy.get('<selector>:invalid')`

```js
cy.get('[data-name="name"]:invalid').should('have.length', 1);
```

2. `validity` property

```js
cy.get('[data-name="name"]').then(($input) => {
  return expect($input[0].validity.tooShort).to.be.true;
});
```

### Desired behavior:

Correctly indicate that the element is indeed invalid (so that assertions can be based on this).

### Test code to reproduce

These tests both fail in showing any problems with invalidity, despite the HTML element being queried having  `minlength` attribute of 3 and our test typing only a single letter ("a"):

```js
it('should catch invalid element by pseudo-selector', () => {
    cy.visit('http://localhost:8000/');
    const tooShortOfAName = 'a';
    cy.get('[data-name="name"]').type(tooShortOfAName);

    // Fails with or without a form click
    // cy.get('[data-name="action2"]').click();

    // Fails here
    cy.get('[data-name="name"]:invalid').should('have.length', 1);
  })

  it('should catch invalid element by JS API', function () {
    cy.visit('http://localhost:8000/');
    const tooShortOfAName = 'a';
    cy.get('[data-name="name"]').type(tooShortOfAName);

    // Fails with or without a form click
    // cy.get('[data-name="action2"]').click();

    // Fails here
    cy.get('[data-name="name"]').then(($input) => {
      return expect($input[0].validity.tooShort).to.be.true;
    });
  });
```

Here's the minimal repo to reproduce: (note that I've added a cypress install as I think you may want to do in your boilerplate also since your docs refer to installing cypress locally)

https://github.com/brettz9/cypress-test-tiny/tree/not-getting-invalidity

(i.e., on the `not-getting-invalidity` branch)

<details>
  <summary>Outdated info</summary>
I've also tacked on an environmental variable test (same as #2605 ?)

```js
it('should obtain environmental variables', function () {
  cy.log(Cypress.env('secret'))
  expect(Cypress.env('secret')).to.equal('Be good');
});
```

...which is not *consistently* getting the `secret` despite it being set in `plugins/index.js`

```js
module.exports = (on, config) => {
  config.env = config.env || {};
  config.env.secret = 'Be good';
  return config;
);
```

...consistent with the approach described for dynamically setting the environmental variables at https://docs.cypress.io/api/plugins/configuration-api.html#Switch-between-multiple-configuration-files or https://docs.cypress.io/api/cypress-api/env.html#From-a-plugin.
</details>

### Versions

4.1.0, Mac OSX 10.15.2, Chrome 80.0.3987.132 (Official Build) (64-bit)

**Comment by brettz9** [comment]
I might add my validity checks are akin to that added in https://github.com/cypress-io/cypress-documentation/pull/1919 (at https://docs.cypress.io/faq/questions/using-cypress-faq.html#Can-I-check-that-a-form%E2%80%99s-HTML-form-validation-is-shown-when-an-input-is-invalid )

**Comment by brettz9** [comment]
My apologies re: the environmental issue; it seems that my issue there may have been not returning the `config` object. However, the main problem with validity checking remains.

**Comment by jennifer-shehane** [comment]
So the problem here is not that it's not finding elements with a psuedo-selector of `invalid`, it's that the actions being performed through Cypress are not triggering the validation checks on the form for some reason. 

You can use jQuery to query for `input:invalid` on the application under test after Cypress performs the type and even the click to submit the form and see that it finds no invalid elements. 

## Reproducible Example

`spec.js`

```js
it('does not find invalid input', () => {
  cy.visit('index.html')
  cy.get('input').type('a')
  // cy.contains('Update').click()
  cy.get('input:invalid').should('have.length', 1)
})
```

`index.html`

```html
<!DOCTYPE html>
<body>
    <input minlength="3">
    <button type="submit">Update</button>
</body>
</html>
```

<img width="730" alt="Screen Shot 2020-03-10 at 2 01 45 PM" src="https://user-images.githubusercontent.com/1271364/76289532-befd8d80-62d7-11ea-9279-b0cafca564fd.png">

Events triggered during type. 

<img width="1166" alt="Screen Shot 2020-03-10 at 1 51 42 PM" src="https://user-images.githubusercontent.com/1271364/76289455-8cec2b80-62d7-11ea-8024-562c11d9546e.png">

If you type manually into the form while the test is running, it will all of a sudden pass. 

<img width="714" alt="Screen Shot 2020-03-10 at 2 04 08 PM" src="https://user-images.githubusercontent.com/1271364/76289662-07b54680-62d8-11ea-8543-5defcd2cb164.png">

**Comment by shamgang** [comment]
In case it helps: I have 2 forms, one with a text and password input, and one with a text, password, and numerical input. In both cases only the text input exhibits this bug. The password and numerical inputs behave as expected - form validation popups are appearing.
Node 12.13.0
Cypress 4.8.0
Chrome 83.0.4103.97
Windows 10 Home v1903 build 18362.900

**Comment by zedrdave** [comment]
Just to add my own experience: getting this bug on a `minlength` validation on a `password` input:
when using `cy.get().type()` to type text that should be too short, I am able to submit the form both manually and through Cypress.

In essence, it looks as if Cypress is not sending the necessary key events for the form to perform validation.

**Comment by shin10** [comment]
I came across this issue as well. The minlength attribute will be ignored. Validation of the pattern
attribute `pattern=".{6}"` however will be triggered. Unfortunately the resulting message ("Please match the requested format") is very vague and rather confusing for the user. Further adding setCustomValidity manually wouldn't be my favourite option.

**Comment by sainthkh** [comment]
I tried [`form.reportValidity()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reportValidity). But somehow, it doesn't work.

**Comment by yanfengliu** [comment]
According to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/minlength), `minLength` is only triggered if value is changed by user.

**Comment by brettz9** [comment]
Is this planned, e.g., on a Roadmap, or given a priority? It is a pretty fundamental need to be able to check forms, testing for validity (for those using the standard mechanisms for doing so). Thanks in advance...

**Comment by sainthkh** [comment]
@brett9 Unfortunately, it's hard to find browser API that triggers form validation. Maybe, this can be solved with native events.

**Comment by brettz9** [comment]
@sainthkh : Thanks for the info and idea. If this would be best addressed by Native Browser Events, I would hope its consideration might be added to the list on #311 , @brian-mann .

**Comment by bahmutov** [comment]
An update: for me the validations are all working, you can see in https://glebbahmutov.com/cypress-examples/6.4.0/recipes/form-validation.html 

Note: `cy.submit` does not check the validity of the form before submitting https://github.com/cypress-io/cypress/issues/14911 while clicking on the submit button does check

**Comment by brettz9** [comment]
Sorry for the delay in replying...

As one example, I have:

```html
<input type="password" autocomplete="new-password" class="form-control" required="" name="pass" id="pass-tf" data-name="reset-pass" minlength="6">
```

but this code fails:

```js
      const tooShortPassword = 'a';
      cy.get('[data-name="reset-pass"]').type(tooShortPassword);

      cy.get('[data-name="reset-pass"]:invalid').should('have.length', 1); // Should be 1, but gives 0
```

The minimal test repo in the original post covers this and the other cases I believe are still occurring.

I.e., it's not just about the submission.

**Comment by brettz9** [comment]
#14965 seems not to actually be covering this issue. My concern is *before* submit, so I think this issue should still be open.

**Comment by jennifer-shehane** [comment]
@brettz9 Yeah, you're right. I'm actually not able to get the original example to work with the new implementation. https://github.com/brettz9/cypress-test-tiny/tree/not-getting-invalidity?rgh-link-date=2020-03-09T04%3A34%3A03Z cc @sainthkh 

1. I don't see `.submit()` failing the test as the PR described it should.
2. I don't see a way to check the conditions of the validity after submitting the form. How do you verify that the correct message is displayed for an invalid field in the example?

**Comment by sainthkh** [comment]
Confirmed that #14965 didn't fix the issue.

But the problem is that it cannot be fixed now. Because all validations are not created equal in browsers. As for the validations like `required`, `mix`, `min`, they always work whether it's a user input or not. But for those like `minlength`, `maxlength`, they are only triggered when users input the value. 

If you want to test it yourself, sample code is below:

```html
<input id="x" minlength="3" value="a" />
<input id="y" required />
```

```js
cy.get('#x').then($x => expect($x[0].checkValidity()).to.be.false) // => fail
cy.get('#y').then($y => expect($y[0].checkValidity()).to.be.false) // => succeed
```

Native events might solve the problem. But I'm not sure about that.

**Comment by tusharAdvanced** [comment]
Is there any solution on this ? I am facing  the same.

**Comment by a-bashkatov** [comment]
I could solve this problem using "realType" from [cypress-real-events](https://github.com/dmtrKovalenko/cypress-real-events),  then the input field is validated as with real text input. However, this solution is not suitable, for example, for Firefox.

**Comment by fifkapetr** [comment]
Issue replicated in version: 14.5.4
Simple form with HTML validations: https://tredgate.com/webtrain/registration.html

Test code
```javascript
describe("Form validation suite", () => {
  it("cy.submit validations", () => {
    cy.visit("https://tredgate.com/webtrain/registration.html");
    cy.get("#myForm").submit();
  });
});
```
